File: jediterm-core-mpp/src/jvmMain/java/com/jediterm/core/Color.java
package com.jediterm.core;

import org.jetbrains.annotations.NotNull;

public final class Color {
  private final int value;

  public Color(int r, int g, int b) {
    this(r, g, b, 255);
  }

  public Color(int r, int g, int b, int a) {
    value = ((a & 0xFF) << 24) | ((r & 0xFF) << 16) | ((g & 0xFF) << 8) | (b & 0xFF);
  }

  public Color(int rgb) {
    this(rgb, false);
  }

  public Color(int rgba, boolean hasAlpha) {
    if (hasAlpha) {
      value = rgba;
    } else {
      value = 0xff000000 | rgba;
    }
  }

  public int getRed() {
    return (value >> 16) & 0xFF;
  }

  public int getGreen() {
    return (value >> 8) & 0xFF;
  }

  public int getBlue() {
    return value & 0xFF;
  }

  public int getAlpha() {
    return (value >> 24) & 0xff;
  }

  public int getRGB() {
    return value;
  }

  /**
   * Returns <a href="https://linux.die.net/man/3/xparsecolor">XParseColor</a> representation of TerminalColor
   */
  public @NotNull String toXParseColor() {
    return "rgb:" + toHexString16();
  }

  private @NotNull String toHexString16() {
    // (n * 0x101) converts the 8-bit number to 16 bits.
    String red = padStart(Integer.toHexString(getRed() * 0x101), 4, '0');
    String green = padStart(Integer.toHexString(getGreen() * 0x101), 4, '0');
    String blue = padStart(Integer.toHexString(getBlue() * 0x101), 4, '0');

    return red + "/" + green + "/" + blue;
  }

  @SuppressWarnings("SameParameterValue")
  private @NotNull String padStart(@NotNull String str, int totalLength, char ch) {
    return Character.toString(ch).repeat(Math.max(0, totalLength - str.length())) + str;
  }

  @Override
  public boolean equals(Object o) {
    return this == o || o instanceof Color && ((Color)o).value == value;
  }

  @Override
  public int hashCode() {
    return value;
  }

  public String toString() {
    return getClass().getName() + "[r=" + getRed() + ",g=" + getGreen() + ",b=" + getBlue() + ", alpha=" + getAlpha() + "]";
  }
}

File: bossterm-core-mpp/src/jvmMain/kotlin/ai/rever/bossterm/core/Color.kt
package ai.rever.bossterm.core

import kotlin.math.max

class Color {
    val rGB: Int

    @JvmOverloads
    constructor(r: Int, g: Int, b: Int, a: Int = 255) {
        this.rGB = ((a and 0xFF) shl 24) or ((r and 0xFF) shl 16) or ((g and 0xFF) shl 8) or (b and 0xFF)
    }

    constructor(rgb: Int) : this(rgb, false)

    constructor(rgba: Int, hasAlpha: Boolean) {
        if (hasAlpha) {
            this.rGB = rgba
        } else {
            this.rGB = -0x1000000 or rgba
        }
    }

    val red: Int
        get() = (this.rGB shr 16) and 0xFF

    val green: Int
        get() = (this.rGB shr 8) and 0xFF

    val blue: Int
        get() = this.rGB and 0xFF

    val alpha: Int
        get() = (this.rGB shr 24) and 0xff

    /**
     * Returns [XParseColor](https://linux.die.net/man/3/xparsecolor) representation of TerminalColor
     */
    fun toXParseColor(): String {
        return "rgb:" + toHexString16()
    }

    private fun toHexString16(): String {
        // (n * 0x101) converts the 8-bit number to 16 bits.
        val red = padStart(Integer.toHexString(this.red * 0x101), 4, '0')
        val green = padStart(Integer.toHexString(this.green * 0x101), 4, '0')
        val blue = padStart(Integer.toHexString(this.blue * 0x101), 4, '0')

        return red + "/" + green + "/" + blue
    }

    private fun padStart(str: String, totalLength: Int, ch: Char): String {
        return ch.toString().repeat(max(0, totalLength - str.length)) + str
    }

    override fun equals(other: Any?): Boolean {
        return this === other || other is Color && other.rGB == this.rGB
    }

    override fun hashCode(): Int {
        return this.rGB
    }

    override fun toString(): String {
        return javaClass.getName() + "[r=" + this.red + ",g=" + this.green + ",b=" + this.blue + ", alpha=" + this.alpha + "]"
    }
}
